Feature/certificate designs

norrec99 opened the pull request:
BEFORE

![image](https://user-images.githubusercontent.com/45902312/153408194-c968a10d-11db-4a2e-9281-2afdebccc69a.png)

AFTER

![image](https://user-images.githubusercontent.com/45902312/153418185-c645290a-ab04-44af-b83d-0cd9f96b3589.png)

ryancwalsh commented on web-app/pages/api/cert/[imageFileName].ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable max-lines-per-function */
```
Please don't disable max-lines-per-function up here at the top, where it would apply to the whole file. And let's ideally not disable it anywhere. But *occasionally* there might be one function that needs ~32 lines or so and doesn't seem like it needs to be split into smaller functions. But that might be rare.

norrec99 replied:
I couldn't make `fetchCertificateDetails` less than 30 lines so I put `// eslint-disable-next-line max-lines-per-function` just above the `fetchCertificateDetails `

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -1,17 +1,13 @@
-const fontFamily = 'signpainter';
+export const width = 1080; // width of the image
+export const height = 1080; // height of the image
+const fontFile = './fonts/Manrope-VariableFont_wght.ttf';
```
If this is the only font file that we'll ever need, then maybe we should just delete the other ttf files from the repo.

norrec99 replied:
done

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -31,23 +27,36 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
+  const programDescription2 = programDescription.split(',')[1];
+  console.log(newDate);
+
+  const grey = '#757575';
```
Please rename "grey" to "gray".

norrec99 replied:
done

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -31,23 +27,36 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
 
-  const fillStyle = `#${Math.floor(Math.random() * 16777215).toString(16)}`; // This is just temporary and will be removed once we have final designs for each program's certificate. https://css-tricks.com/snippets/javascript/random-hex-color/
-  const font = `40px '${fontFamily}' bold`;
+  const newDate = date.split(' ')[0];
```
In general, I try to avoid hard-coding things like `[0]` and `[1]`. For this particular line, if what you're doing is just trying to change the appearance of a date, search the repo for `dayjs`. Format the date with that kind of helper.

norrec99 replied:
done

norrec99 commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -31,23 +27,36 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
+  const newDate = date.split(' ')[0];
+  const newExpiration = expiration.split(' ')[0];
+  const programDescription1 = programDescription.split(',')[0];
+  const programDescription2 = programDescription.split(',')[1];
```
I tried to explain why i did such a thing in here: https://discord.com/channels/828768337978195969/906115083250307103/941290874573320203

But basically, if I don't split the description, it looks like this
![image](https://user-images.githubusercontent.com/45902312/153601833-ce07d249-c976-4e94-8d38-8fdaeb0b7f5b.png)

After splitting it,
![image](https://user-images.githubusercontent.com/45902312/153601926-8d3243f5-33b8-4638-9ea7-d0369bb9a305.png)

Maybe it is not a best practice but idk how to fix it otherwise. 
So for now, I can not change it.

ryancwalsh commented:
We're removing that field (competencies/description) anyway. Dan is
updating the designs to remove it, right?

On Fri, Feb 11, 2022 at 8:44 AM Cagatay Soyer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In web-app/helpers/certificate-designs.ts
> <https://github.com/NEAR-Edu/near-certification-tools/pull/5#discussion_r804662650>
> :
>
> >
> -  const fillStyle = `#${Math.floor(Math.random() * 16777215).toString(16)}`; // This is just temporary and will be removed once we have final designs for each program's certificate. https://css-tricks.com/snippets/javascript/random-hex-color/
> -  const font = `40px '${fontFamily}' bold`;
> +  const newDate = date.split(' ')[0];
> +  const newExpiration = expiration.split(' ')[0];
> +  const programDescription1 = programDescription.split(',')[0];
> +  const programDescription2 = programDescription.split(',')[1];
>
> I tried to explain why i did such a thing in here:
> https://discord.com/channels/828768337978195969/906115083250307103/941290874573320203
>
> But basically, if I don't split the description, it looks like this
> [image: image]
> <https://user-images.githubusercontent.com/45902312/153601833-ce07d249-c976-4e94-8d38-8fdaeb0b7f5b.png>
>
> After splitting it,
> [image: image]
> <https://user-images.githubusercontent.com/45902312/153601926-8d3243f5-33b8-4638-9ea7-d0369bb9a305.png>
>
> Maybe it is not a best practice but idk how to fix it otherwise.
> So for now, I can not change it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NEAR-Edu/near-certification-tools/pull/5#discussion_r804662650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP5MXLBAYIJF3SZDEDHJJTU2UHFHANCNFSM5OARVQFA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

norrec99 commented:
I never heard that Dan is working on removing the "description" from the design anywhere. Where did you see it? Maybe I missed something.

norrec99 commented:
[fb76781](https://github.com/NEAR-Edu/near-certification-tools/pull/5/commits/fb7678188a80712cebc2a41a627f75725ea84bde)
[e94162b](https://github.com/NEAR-Edu/near-certification-tools/pull/5/commits/e94162bc8822bc1a3d4c76eb11f18408cb7b5b75)
For the latest commits, here are the changes: 

I don't have the latest svgs yet so this is just a placeholder. As soon as I get all the svgs from Dan, It will be ready for production. 
Before
![Untitled](https://user-images.githubusercontent.com/45902312/153622941-6db52168-828f-4bde-a61a-abbc87fa8ee6.png)

After
![image](https://user-images.githubusercontent.com/45902312/153622122-98affbca-3015-4079-a3d6-282833c97d56.png)

So, I deleted the `populateAnalystCert` function because we are now using `switch`  to generate different certs. 

I also deleted the `switch` in the `web-app/pages/api/cert/[imageFileName].ts` for the same reason above.

ryancwalsh commented:
The Discord near-certification-tools channel.

Sherif said he doesn't want competencies. I'm pretty sure he means this
field.

On Fri, Feb 11, 2022, 9:02 AM Cagatay Soyer ***@***.***>
wrote:

> I never heard that Dan is working on removing the "description" from the
> design anywhere. Where did you see it? Maybe I missed something.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1036243714>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP5MXKBUBZOR2URG67IDR3U2UJF7ANCNFSM5OARVQFA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

norrec99 commented:
With the last commits, the page looks like this

![image](https://user-images.githubusercontent.com/45902312/153925596-1aa96ddd-563a-4f06-9796-88074fa0f6bb.png)

ryancwalsh commented:
@norrec99 Thanks for that helpful screenshot. It seems that the certificates have a blue border, so I wonder if we should remove the light gray border surrounding each of those, since it seems potentially unnecessary.

Currently, http://localhost:3000/account/hatchet.testnet throws this error for me:

```
error - (api)/helpers/certificate-designs.ts (38:50) @ populateDeveloperCert
TypeError: Cannot read properties of undefined (reading 'split')
  36 |   const programDescription4 = programDescription.split('program ')[1];
  37 |   const programDescription1 = programDescription.split('by')[0];
> 38 |   const programDescription2 = programDescription4.split('on ')[0];
     |                                                  ^
  39 |   const programDescription3 = programDescription4.split('contracts ')[1];
  40 | 
  41 |   const gray = '#757575';
```

and looks like this: 
![image](https://user-images.githubusercontent.com/2086493/153928167-a735d590-6dcc-4e5c-9330-89f408d37880.png)

So, this helps show that you'll need to handle that programDescription section differently. You'll need to figure out a way other than splitting it based on hard-coded text values that you were searching for, since you can't guarantee that those pieces of text will exist.

I found https://stackoverflow.com/questions/4991171/auto-line-wrapping-in-svg-text

But embedding HTML doesn't seem like a good idea since we want these images to be shareable as PNGs too (on Twitter, etc).

So, you can explore whether you can come up with a solution.  If you feel blocked, then maybe you should explain to Dan, Sherif, me, and others in the Discord channel and recommend temporarily removing these multi-line fields.  And then if we can ever figure out a graceful way of adding a multi-line description to an SVG, we can add it later. 

What do you think?

ryancwalsh commented:
@norrec99 Yeah, we don't want our code to be brittle and break / throw an error just because data changes underneath. I.e. we want the pages / images to look reasonable regardless of what the on-chain data says, ideally.

When I wrote _"But embedding HTML doesn't seem like a good idea since we want these images to be shareable as PNGs too (on Twitter, etc)."_ I was referring to the workarounds suggested in the StackOverflow answers. They suggested using embedded HTML as a way to show multi-line text. That might be fine if we only ever wanted to display the SVGs on our webpage. But in reality, we want the SVGs to be exportable and also downloadable as PNGs. If we were just displaying a webpage, we would just use HTML and CSS from the beginning and never bother with SVG at all. So that workaround suggestion in those S.O. answers is kind of bizarre.

norrec99 commented:
@ryancwalsh I'm looking for a way to solve the `programDescription` but so far I couldn't find a way to do it. 

I found this https://stackoverflow.com/questions/16701522/how-to-linebreak-an-svg-text-within-javascript

They are talking about adding something like this;
```js
<tspan x="0" dy="1.2em">very long text</tspan>
<tspan x="0" dy="1.2em">I would like to linebreak</tspan>
``` 

But even in this case, I guess we need to use `.split()` and put those into the `<tspan>` which is not the behavior we want. 

I'll research a bit more and if I can't find anything useful, I will ask for help on the `near-certification-tools` channel.

norrec99 commented:
The other issue I want to mention is the alignment of the elements

Here you can see the program name is close to the right-hand side because the text is too long (relatively) 
![image](https://user-images.githubusercontent.com/45902312/154046222-f62e916b-b453-4de3-8561-f3c4da2710d5.png)

On the other hand, this is close to the left-hand side because the text is too short (relatively) 
![image](https://user-images.githubusercontent.com/45902312/154046512-14602975-f139-4102-bb98-4ad931b4f36d.png)

This issue also applies to the `accountName` section and probably to the `instructor` too. If those texts will be too long or too short then the alignment will be broken because we are applying the same width and height to all.

Is there a way to prevent such behavior?

ryancwalsh commented:
@norrec99 See https://duckduckgo.com/?q=canvas+how+center+align+text&t=brave&ia=web

Also I'm curious whether https://stackoverflow.com/a/28553412/470749 helps the multiline problem.

norrec99 commented:
I think we find a way to this multiline problem. Thanks to @rashaabdulrazzak we come up with a function that splits long text into shorter lines.

There are a couple of questions and comments on the latest commit as I wrote them as comments but the major problem is this
![image](https://user-images.githubusercontent.com/45902312/154078430-bcb2715a-41ab-4e28-9a7b-c3f6cdb7ffc8.png)
There is an extra space that I couldn't remove at the beginning of the text.

Also, font-weight doesn't really work at all! That's another issue.

ryancwalsh commented:
Did you search for "how to remove a space from the beginning of a string"?

Also, if /fonts/Manrope-VariableFont_wght.ttf doesn't support ExtraBold, and if you need ExtraBold, then it sounds like you'll need an additional font file for that weight.

Also, since Dan chose not to use a monospaced font for the dates in the bottom right, and since they aren't right-aligned, they end up looking wonky. I think those should be right-aligned. And maybe that will also mean that we should remove the labels "Issued:" and "Expiring:" from the background image and dynamically insert them into the right-aligned text.

The fonts for the instructor (label and value) also look weird currently.

You're making progress, though!

norrec99 commented:
I just fixed the leading space by using `trim()`

> Also, if /fonts/Manrope-VariableFont_wght.ttf doesn't support ExtraBold, and if you need ExtraBold, then it sounds like you'll need an additional font file for that weight.

I don't think it is the reason because "medium", "regular" or "bold" are not working either. Even though I put `Manrope-Bold.tff`, doesn't seem to be working.

I tried to change the order of this.  ```const accountFont = `60px '${fontFamily}' bold`;```
For example like that ```const accountFont = `bold 60px '${fontFamily}'`;```
Still, nothing changed

> And maybe that will also mean that we should remove the labels "Issued:" and "Expiring:" from the background image and dynamically insert them into the right-aligned text.

If we are going to do this, we should also do it for the instructor I guess because it looks weird as you said.

> You're making progress, though!

Thanks!

ryancwalsh commented:
The instructor field won't have an alignment problem, though. So the problem there is just that we should adjust the font style and/or size of either the label or the value.

@rashaabdulrazzak Do you think you could help figure out how to get the fonts to look good? For example, figure out why bold isn't working. Thanks.

norrec99 commented:
@ryancwalsh I fixed the font-weight problem. 

![image](https://user-images.githubusercontent.com/45902312/154220883-d24bb39d-83a6-4279-9ec6-4dbaa8eb28e9.png)


> The instructor field won't have an alignment problem, though. So the problem there is just that we should adjust the font style and/or size of either the label or the value.

Do you still think the "instructor" field looks weird?

cc @rashaabdulrazzak

norrec99 commented:
> Also, since Dan chose not to use a monospaced font for the dates in the bottom right, and since they aren't right-aligned, they end up looking wonky. I think those should be right-aligned. And maybe that will also mean that we should remove the labels "Issued:" and "Expiring:" from the background image and dynamically insert them into the right-aligned text.

@ryancwalsh So, I was experimenting with the date field and realized something minor. 
For example, think about this two date
 "2021-11-28"
 "2022-08-16"
I put the dates exactly on the same x-axis above on purpose and they look like they have the same width. But this scenario won't work if the font Github uses here would be the "manrope" that we use in our project. It's because "1" has a smaller width than "2" or "0" has bigger width than "1" in the "manrope". That's why we see an alignment problem in the date fields on the cert. Even though we make them right-aligned, we have a high chance to encounter an alignment problem. 

See the differences below;
![image](https://user-images.githubusercontent.com/45902312/154225669-9cbe90ab-478a-45c3-b72b-3345f399c32b.png)
---
![image](https://user-images.githubusercontent.com/45902312/154225846-dd4bc8c9-015c-47aa-a146-88f95e4659a0.png)

wdyt?

norrec99 commented:
What do you think about these designs? @ryancwalsh @rashaabdulrazzak @ozanisgor 

With many certs
![image](https://user-images.githubusercontent.com/45902312/154237522-25c62c99-69be-425b-bdc1-1f3757b9d317.png)
With one cert
![image](https://user-images.githubusercontent.com/45902312/154238539-b6052c1d-8dcb-42e3-9ffb-c9113d0df281.png)
---

With many certs
![image](https://user-images.githubusercontent.com/45902312/154237688-628917e2-c5e2-4742-a721-ebf66d95e775.png)
With one cert
![image](https://user-images.githubusercontent.com/45902312/154238690-76fedef5-ace4-4c6e-a0b3-db4db2d9a37d.png)

---
This is the current one
With many certs
![image](https://user-images.githubusercontent.com/45902312/154237749-beeb4631-68b3-4236-a7b1-cfbdd8759ef2.png)
With one cert
![image](https://user-images.githubusercontent.com/45902312/154239722-41733652-f5c1-472a-85e1-f157fe175cd2.png)

ozanisgor commented:
> What do you think about these designs? @ryancwalsh @rashaabdulrazzak @ozanisgor

I like the first one but If person will have only one certificate It will not look good when aligned to left. If there is a way to make cert aligned to center when there is only one then my choice is the first alternative that you share but otherwise current version is also looking good.

norrec99 commented:
@ryancwalsh I was working on this problem https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1040127826

To fix this issue, I changed the structure a little. If you remember when I asked, "do we need to create a function for each cert to generate an image?" and you said, "we can use a switch, if all the properties are the same for each cert". Then we decided to go with the "switch solution".
So I literally create a function for each cert in order to play them separately. Now we can change the position of each element on the cert.

Here is the look of the page
![image](https://user-images.githubusercontent.com/45902312/154262842-71f9f057-c30a-45ed-b66c-219d767c6314.png)

I haven't committed it yet. What do you think about this solution?

ryancwalsh commented:
@norrec99 Thanks for the screenshots and all of these comments. I don't know what you're saying in the most recent comment, though, because you didn't push any commits. If you're nervous about messing with your current branch, you could always just branch off of that branch and push commits there for me to see. Then we can abandon and delete whichever branch we don't like.

The new font family that you got from Dan is what I was talking about when I said "monospace". Monospace fonts take the same width for each character, unlike most fonts. It makes design alignment much easier but is less pretty.  I think we can proceed with monospace in the short run. Later, we can review the other suggestion I was making since I don't think you understood, and we can see what people prefer.

@ozanisgor and @rashaabdulrazzak you should hear this too: pushing commits to GitHub is not the same thing as then also submitting a PR on GitHub. We should only submit a PR when we think something is ready to be reviewed and deployed. We've all been a little sloppy about this recently. And we've left these PRs open rather than changing them into "draft" mode or canceling them. 

E.g. you know that I wouldn't approve this PR currently because this is the result I see: 
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/2086493/154279778-1ebfeb50-6162-4ca7-b654-e67ba318275e.png">

In general, whenever possible, it's best to work iteratively. It's best for the reviewer (and the rest of the team too) if you make solid progress on tiny pieces and make those pieces production-quality, and then submit a PR for that, and that makes it easy for me to review it, approve it, and merge it into `develop` for everyone else. 

This means that then we'll have fewer conflicts and also our server will show improvements more frequently.

ryancwalsh commented:
@eadsoy @hiba-machfej It's worth subscribing to all the conversations of this repo if you haven't already so that you can follow along some tips I'm sharing.

ryancwalsh commented:
For example, @norrec99 when I saw the screenshot in the PS part of your comment at https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1041239687 I thought "That looks like a good enough result for right now that it would be worth deploying". But the current state of the commits pushed to this branch (i.e. this PR) aren't what's shown in that screenshot. So I can't approve and deploy it.

My general approach is that I always lean towards approving (if something is above a minimum threshold of quality) even if I have ideas for improving it further later in a next iteration.

norrec99 commented:
> E.g. you know that I wouldn't approve this PR currently because this is the result I see:

I don't understand why are you seeing something like that. 
You should have seen this image with the latest commit
![image](https://user-images.githubusercontent.com/45902312/154287309-53bf91a4-05d0-4afe-b088-5ec8a2565aab.png)

Maybe you should use my dev-account: dev-1644856769338-22652649644211
idk

I'll try to replicate the error you got on this post https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1041521643

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -1,17 +1,16 @@
+export const height = 1080; // height of the image
+const fontFile = './fonts/Manrope-VariableFont_wght.ttf';
+const fontFamily = 'Manrope, Sans Serif';
+const fontFile2 = './fonts/DMMono-Medium.ttf';
```
@norrec99 Please don't append numbers to variable names. Please rename each variable to make them more obvious about what each is referring to.

norrec99 replied:
Done

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -28,29 +27,105 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
+  switch (programCode) {
+    case 'TR101': // programCode needs to change in a meaningful way.
+      image = await loadImage(certificateBackgroundNcdImage);
+      console.log({ image });
```
You have `console.log({ image });` repeated many times in this function. Please move it below the switch if it achieves what you're trying to do.

ryancwalsh commented on web-app/pages/index.tsx:
```diff
@@ -18,16 +18,30 @@ const Home: NextPage = () => {
-          Go
-        </button>
-      </form>
+      <div className="">
```
All of these changes here in this diff in index.tsx seem like the same problem @ozanisgor is having. It seems like you haven't rebased with the develop branch. These changes already happened in develop. So if *you* didn't work on index.tsx for your feature, there should be no diff (changes) shown here.

norrec99 replied:
I think all of us need a live tutorial about how to do a clear rebase from you.

cc @ozanisgor @rashaabdulrazzak

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -28,29 +27,105 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
+// eslint-disable-next-line max-lines-per-function
 export async function populateDeveloperCert(canvas: Canvas, details: any) {
-  // TODO
   console.log('populateDeveloperCert', { details });
```
You need to update this function name (including the string in the console log here), right? Because it's no longer just about the Developer program.

norrec99 replied:
done

ryancwalsh commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -28,29 +27,105 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
   context.fillText(text, leftPos, rightPos);
 }
 
+// split long text into shorter lines
```
Consider using JSDoc comment formatting here when stating what a function does. This will help IDE hints.

norrec99 replied:
Done

ryancwalsh commented:
@norrec99 I think maybe the problem I'm experiencing locally is related to the fact that you have 2 files that are basically the same file name but with different capitalization. `web-app/public/certificate-backgrounds/NCD_certificate.svg`

Different systems treat capitalization differently. My Mac seems not to like it.  I don't know the solution yet. But I think you'll probably need to remove one from the repo.

By the way, I know I'm giving you a lot more work to do in these suggestions of mine. Maybe @rashaabdulrazzak or @ozanisgor can work with you if they have capacity and have finished their stuff already.

norrec99 commented:
This is awkward because I deleted the file with lower letters a long time ago. In this commit to be specific: https://github.com/NEAR-Edu/near-certification-tools/pull/5/commits/a2b8874786f5e27ce708a0dd0f90b817f685b519
Now I only have these files:
![image](https://user-images.githubusercontent.com/45902312/154298273-2ae36135-8dfa-4e02-a857-6d7279eb1f40.png)

> By the way, I know I'm giving you a lot more work to do in these suggestions of mine. Maybe @rashaabdulrazzak or @ozanisgor can work with you if they have capacity and have finished their stuff already.

We are already working together, I ask lots of questions to them and we make calls to discuss those questions. But please put something more on my shoulders, this would only make me happier :)

ryancwalsh reviewed (approved):
I fixed the lowercase file situation. I'll approve this PR and merge into develop and delete this branch.

But then I want you to create a new branch from develop where you finish working on everything we have been discussing in this PR.

ryancwalsh commented:
@norrec99  I think the alignment of this part is messed up:

<img width="369" alt="image" src="https://user-images.githubusercontent.com/2086493/154320774-3bd5a15f-e0f4-43e8-8319-95d9d4be362b.png">

ryancwalsh commented:
We're making progress! Here is what `main` (production server) looks like now: https://near-certification-tools-tpq1.onrender.com/account/hatchet.testnet

@norrec99  I think the alignment of this part is messed up:

<img width="369" alt="image" src="https://user-images.githubusercontent.com/2086493/154320774-3bd5a15f-e0f4-43e8-8319-95d9d4be362b.png">

norrec99 commented:
Yes, I'm aware of this issue. Actually, I was trying to explain this alignment issue in this post https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1041431054

The image you see on that post is the one that solves the alignment problem by separating each cert with a different function. 

This is the one I didn't commit. I will commit this change and all the other feedback you mentioned tomorrow.

ryancwalsh commented:
Ok, but I don't understand why it would be anything specific to each
program. Instead, we just want *all* program names centered. I would have
thought that would be simple but I guess I'm misunderstanding. 😃

On Wed, Feb 16, 2022, 10:42 AM Cagatay Soyer ***@***.***>
wrote:

> Yes, I'm aware of this issue. Actually, I was trying to explain this
> alignment issue in this post #5 (comment)
> <https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1041431054>
>
> The image you see on that post is the one that solves the alignment
> problem by separating each cert with a different function.
>
> This is the one I didn't commit. I will commit this change and all the
> other feedback you mentioned tomorrow.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NEAR-Edu/near-certification-tools/pull/5#issuecomment-1041923153>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP5MXMQB4OPWYKIZGKSFLTU3POX3ANCNFSM5OARVQFA>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

norrec99 commented:
I think it is because of this function

'addText(canvas, programName, programFont, black, 240, 680);' 

We are saying that no matter what every programName starts in the 240px at the x-axis. But some programNames are longner and some are shorter. This causes the alignment problem.

Maybe there is a better way than hard-coding the x-axis but I decided to do a different function for each cert so that I can change 

'addText(canvas, programName, programFont, black, 240, 680);' 

this function for each cert.

ryancwalsh commented:
@norrec99  Thanks for that explanation. The better approach here is not to hard-code X position values based on what we assume the program names will be. Instead we want the X pos value that we pass to addText to be talking about the left position of the "div" (or equivalent). Then we also want (maybe via a new argument) to be able to do the equivalent of text-align center in certain cases, such as for the program name field.

https://github.com/NEAR-Edu/near-certification-tools/blob/develop/web-app/helpers/certificate-designs.ts#L22

norrec99 commented on web-app/helpers/certificate-designs.ts:
```diff
@@ -28,29 +27,105 @@ function addText(canvas: Canvas, text: string, font: string, fillStyle: string,
+
+    // doesn't really understand the condition
+    // ask Ryan about it
+    if (testWidth > maxWidth && n > 6) {
```
Thanks! I created a const for it

norrec99 commented:
@ryancwalsh What do you think of this solution for the alignment problem? 4bab2ad4b09460329c9e123f4b2b99888cdd6e05

Changed text alignment from "left" to "center" except for the `programDescription` 
Made some calculations for the x-axis in the `addText()` function. It might be confusing but it works for now 😅

Page looks like this:
![image](https://user-images.githubusercontent.com/45902312/154484588-866c975f-078e-4a0d-a46f-a5f3d91b8ee1.png)

cc @ozanisgor @rashaabdulrazzak
